src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                        diagnostics.Add(ErrorCode.ERR_NonTaskMainCantBeAsync, candidate.Locations.First(), candidate);
+                        continue;
                     }
-
                     viableEntryPoints.Add(candidate);
                 }
 
-                if ((object)mainType == null || viableEntryPoints.Count == 0)
+                if (viableEntryPoints.Count == 0)
+                {
+                    foreach (var (IsValid, Candidate, SpecificDiagnostics) in taskEntryPoints)
+                    {
+                        if (Candidate.IsGenericMethod || Candidate.ContainingType.IsGenericType)
+                        {
+                            // a single error for partial methods:
+                            noMainFoundDiagnostics.Add(ErrorCode.WRN_MainCantBeGeneric, Candidate.Locations.First(), Candidate);
+                            continue;
+                        }
+
+                        if (!IsValid)
+                        {
+                            noMainFoundDiagnostics.Add(ErrorCode.WRN_InvalidMainSig, Candidate.Locations.First(), Candidate);
+                            noMainFoundDiagnostics.AddRange(SpecificDiagnostics);
+                            continue;
+                        }
+
+                        // PROTOTYPE(async-main): Get the diagnostic to point to a smaller syntax piece.
+                        if (CheckFeatureAvailability(Candidate.GetNonNullSyntaxNode(), MessageID.IDS_FeatureAsyncMain, diagnostics))
+                        {
+                            diagnostics.AddRange(SpecificDiagnostics);
+                            viableEntryPoints.Add(Candidate);
+                        }
+
+                    }
+                }
+
+                foreach (var (_, _, SpecificDiagnostics) in taskEntryPoints)
                 {
-                    diagnostics.AddRange(warnings);
+                    SpecificDiagnostics.Free();
                 }
 
-                warnings.Free();
+                if (viableEntryPoints.Count == 0)
+                {
+                    diagnostics.AddRange(noMainFoundDiagnostics);
+                } 
+                else if ((object)mainType == null)
+                {
+                    // Only add warning diagnostics.  The reason that Error diagnostics can end up in `noMainFoundDiagnostics` is when 
+                    // HasEntryPointSignature yields some Error Diagnostics when people implement Task or Task<T> incorrectly.
+                    //
+                    // We can't add those Errors to the general diagnostics bag because it would break previously-working programs.
+                    // The fact that these warnings are not added when csc is invoked with /main is possibly a bug, and is tracked at
+                    // https://github.com/dotnet/roslyn/issues/18964
+                    diagnostics.AddRange(noMainFoundDiagnostics.AsEnumerable().Where(d => d.Severity == DiagnosticSeverity.Warning));
```

Comment:
>noMainFoundDiagnostics.AsEnumerable().Where(d => d.Severity == DiagnosticSeverity.Warning) [](start = 41, length = 90)

Linq is banned in compilers. Also, if the point of this code is to preserve old behavior, then we should probably be more precise and add only  WRN_MainCantBeGeneric and WRN_InvalidMainSig warnings.